src/types.jl

```diff
@@ -1,85 +1,74 @@
-#-------------------
-#--------------------
-#### Symbolic
-#--------------------
 abstract type Symbolic{T} end
 
-###
-### Uni-type design
-###
-
-@enum ExprType::UInt8  SYM TERM ADD MUL POW DIV
+@enum ExprType::UInt8 SYM TERM ADD MUL POW DIV CONST
 
-const Metadata = Union{Nothing,Base.ImmutableDict{DataType,Any}}
+const Metadata = Union{Nothing, Base.ImmutableDict{DataType, Any}}
 const NO_METADATA = nothing
+const EMPTY_HASH = UInt(0)
 
-sdict(kv...) = Dict{Any, Any}(kv...)
-
-using Base: RefValue
-const EMPTY_ARGS = []
-const EMPTY_HASH = RefValue(UInt(0))
-const NOT_SORTED = RefValue(false)
-const EMPTY_DICT = sdict()
-const EMPTY_DICT_T = typeof(EMPTY_DICT)
-
-@compactify show_methods=false begin
-    @abstract struct BasicSymbolic{T} <: Symbolic{T}
-        metadata::Metadata     = NO_METADATA
+@adt BasicSymbolicImpl begin
+    struct Sym
+        name::Symbol
     end
-    struct Sym{T} <: BasicSymbolic{T}
-        name::Symbol           = :OOF
+    struct Term
+        f::Any
+        arguments::Vector{BasicSymbolic}
     end
-    struct Term{T} <: BasicSymbolic{T}
-        f::Any                 = identity  # base/num if Pow; issorted if Add/Dict
-        arguments::Vector{Any} = EMPTY_ARGS
-        hash::RefValue{UInt}   = EMPTY_HASH
+    struct Add
+        coeff::BasicSymbolic
```

Comment:
@YingboMa Should we restrict the type of the `coeff` field in `Add` and `Mul` to `Const` here?